core/services/pipeline/runner.go

```diff
@@ -236,133 +300,94 @@ func (r *runner) executeRun(
+	pipeline, err := Parse([]byte(spec.DotDagSource))
 	if err != nil {
 		return run, nil, false, err
 	}
 
-	headTaskRuns, err := r.memoryTaskRunDAGFromTaskDAG(taskDAG, pipelineInput, txdb)
-	if err != nil {
-		return run, nil, false, err
+	// initialize certain task params
+	txMu := &sync.Mutex{}
+	for _, task := range pipeline.Tasks {
+		if task.Type() == TaskTypeHTTP {
+			task.(*HTTPTask).config = r.config
+		} else if task.Type() == TaskTypeBridge {
+			task.(*BridgeTask).config = r.config
+			task.(*BridgeTask).safeTx = SafeTx{txdb, txMu}
+		}
 	}
 
+	scheduler := newScheduler(pipeline, pipelineInput)
+
 	// TODO: Test with multiple and single null successor IDs
 	// https://www.pivotaltracker.com/story/show/176557536
 
-	// We execute tasks using "fan in" job processing because tasks have only one
-	// output but potentially many inputs. Start one goroutine per "head task",
-	// having it proceed down its chain of successors until:
-	//  - it encounters a task with unfulfilled predecessors
-	//  - it encounters a task that has already been claimed/finished
-	//  - it processes the final task
-
 	var (
-		vars             = NewVarsFrom(map[string]interface{}{"input": pipelineInput})
-		taskRunResults   TaskRunResults
-		taskRunResultsMu sync.Mutex
-		wg               sync.WaitGroup
-		retry            bool
+		vars  = NewVarsFrom(map[string]interface{}{"input": pipelineInput})
+		retry bool
 	)
-	wg.Add(len(headTaskRuns))
-	for _, mtr := range headTaskRuns {
+
+	for taskRun := range scheduler.taskCh {
+		// execute
 		go func(taskRun *memoryTaskRun) {
 			defer func() {
 				if err := recover(); err != nil {
 					logger.Default.Errorw("goroutine panicked executing run", "panic", err, "stacktrace", string(debug.Stack()))
+
 					// No mutex needed: if any goroutine panics, we retry the run.
 					retry = true
-				}
-				wg.Done()
-			}()
-			for taskRun != nil {
-				// If this one is still waiting another chain, abandon this
-				// goroutine and let the other handle it
-				if taskRun.numPredecessorsRemaining() > 0 {
-					return
-				}
-				// Avoid double execution, only one goroutine may finish the task
-				if ok := taskRun.tryToClaim(); !ok {
-					return
-				}
-
-				taskRunResult := r.executeTaskRun(ctx, spec, vars, taskRun, meta, l)
 
-				if taskRunResult.Result.Error != nil {
-					vars.Set(taskRunResult.Task.DotID(), taskRunResult.Result.Error)
-				} else {
-					vars.Set(taskRunResult.Task.DotID(), taskRunResult.Result.Value)
+					scheduler.resultCh <- TaskRunResult{
+						Task:       taskRun.task,
+						Result:     Result{Error: panicError{err}},
+						FinishedAt: time.Now(),
+						// TODO: CreatedAt
+					}
 				}
+			}()
+			result := r.executeTaskRun(ctx, spec, vars, taskRun, meta, l)
 
-				taskRunResultsMu.Lock()
-				taskRunResults = append(taskRunResults, taskRunResult)
-				taskRunResultsMu.Unlock()
+			// TODO: remove vars locks by constructing Vars instances per task?
```

Comment:
What benefit would this yield?

Reply:
Less locking, `Vars` is just a glorified map with a mutex right now. Since all coordination happens in the scheduler, we can simply give each task a copy of the env at the time of execution.